lib/controllers/responses.js

```diff
@@ -429,53 +388,100 @@ exports.patch = function patch(req, res) {
 
 
 exports.del = function del(req, res) {
-  Response.findOne({
+  Response.findOneAndUpdateAsync({
     'properties.survey': req.params.surveyId,
     'entries._id': req.params.responseId
-  })
-  .exec(function (error, doc) {
-    if (util.handleError(error, res)) { return; }
-    if (doc === null) {
+  }, {
+    $pull: {
+      entries: {
+        _id: req.params.responseId
+      }
+    }
+  }, {
+    new: false
+  }).then(function (doc) {
+    if (!doc) {
       res.send(404);
       return;
     }
 
-    if (doc.entries.length === 1) {
-      // If there's just one entry, remove the whole parent object.
-      doc.remove(function (error) {
-        if (util.handleError(error, res)) { return; }
-        res.send(204);
+    doc = new Response(doc);
```

Comment:
It looks like Mongoose's update-related functions are much more raw than the find/save ones. We also have to explicitly apply pre-save middleware, for example. It makes some sense, but it can make things a little messy.